public-api/src/main/java/org/pmiops/workbench/interceptors/AuthInterceptor.java

```diff
@@ -0,0 +1,114 @@
+import org.springframework.web.method.HandlerMethod;
+import org.springframework.web.servlet.ModelAndView;
+import org.springframework.web.servlet.handler.HandlerInterceptorAdapter;
+
+
+/**
+ * Intercepts all non-OPTIONS API requests to ensure they have an appropriate auth token.
+ *
+ * Checks handler methods for annotations like
+ *     @AuthorityRequired({Authority.REVIEW_RESEARCH_PURPOSE})
+ * to enforce granular permissions.
+ */
+@Service
+public class AuthInterceptor extends HandlerInterceptorAdapter {
+  private static final Logger log = Logger.getLogger(AuthInterceptor.class.getName());
+  private static final String authName = "aou_oauth";
+
+  private final Provider<WorkbenchConfig> configProvider;
+
+  private final PrivateWorkbenchService privateWorkbenchService;
+
+
+  @Autowired
+  public AuthInterceptor(Provider<WorkbenchConfig> configProvider,
+                         PrivateWorkbenchService privateWorkbenchService) {
+    this.configProvider = configProvider;
+    this.privateWorkbenchService = privateWorkbenchService;
+  }
+
+  /**
+   * Returns true iff the request is auth'd and should proceed. Publishes authenticated user info
+   * using Spring's SecurityContext.
+   * @param handler The Swagger-generated ApiController. It contains our handler as a private
+   *     delegate.
+   */
+  @Override
+  public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
+      throws Exception {
+    // Clear the security context before we start, to make sure we're not using authentication
+    // from a previous request.
+    SecurityContextHolder.clearContext();
+
+    // OPTIONS methods requests don't need authorization.
+    if (request.getMethod().equals(HttpMethods.OPTIONS)) {
+      return true;
+    }
+
+    HandlerMethod method = (HandlerMethod) handler;
+
+    boolean isAuthRequired = false;
+    ApiOperation apiOp = AnnotationUtils.findAnnotation(method.getMethod(), ApiOperation.class);
+    if (apiOp != null) {
+      for (Authorization auth : apiOp.authorizations()) {
+        if (auth.value().equals(authName)) {
+          isAuthRequired = true;
+          break;
+        }
+      }
+    }
+    if (!isAuthRequired) {
+      return true;
+    }
+
+    String authorizationHeader = request.getHeader(HttpHeaders.AUTHORIZATION);
+
+    if (authorizationHeader == null || !authorizationHeader.startsWith("Bearer ")) {
+      log.warning("No bearer token found in request");
+      response.sendError(HttpServletResponse.SC_UNAUTHORIZED);
+      return false;
+    }
+
+    String token = authorizationHeader.substring("Bearer".length()).trim();
+
+    if (token.equals("null")) {
+      throw new RuntimeException();
+    }
+
+    Profile profile = privateWorkbenchService.getMe(token);
+    if (configProvider.get().firecloud.enforceRegistered) {
+      if (profile.getIdVerificationStatus() != IdVerificationStatus.VERIFIED) {
+        throw new RuntimeException();
```

Comment:
Probably meant as a placeholder, but should return UnauthorizedException with a recognizable error message we can pick out on the client if needed.

Reply:
So, I agree with adding a recognizeable error message. There is currently no structure in place for custom exceptions in the data browser, so if we want to do that there is some pretty large scope creep. I am not against that, but we should be aware of that if we want to do it.

Reply:
Talked to Karthik. I opened this ticket to handle exceptions in reasonable ways:
`https://precisionmedicineinitiative.atlassian.net/secure/RapidBoard.jspa?rapidView=22&projectKey=DB&view=planning&selectedIssue=DB-120`

Reply:
Should be easy enough to copy the approach used on line 75 of this file. Didn't mean to fixate on specifically using a custom exception type which might not be setup in Data Browser, I just don't want to be returning 500 errors in an expected/common case.

Reply:
Oh! I missed that. Will do.